Allow impl for LocationProvider for to be created during runtime

mickjermsurawong-stripe opened the pull request:
This PR allows users to supply implementation of `LocationProvider` to be created during runtime.
This is motivated from the proposal to solve data localization: https://docs.google.com/document/d/1ZluOiRZlmsfNnQJLSqTiBQg7-XeSE-gvEOn2e0y6E54
From the community sync, the feedback is to make to leave implementation to users so other use cases outside data localization could make use of this too: https://lists.apache.org/list.html?dev@iceberg.apache.org

jackye1995 commented:
I have 2 concerns with this PR:

1. I think we should in addition propose dynamically loading the `LocationProvider` implementation, instead of using a boolean to switch on or off each implementation. This will allow any user to easily plugin a custom implementation. I wonder if such proposal is enough to both satisfy your need and also keep the community version general.

2. for this specific implementation, the approach of using table property to get `localizedDataLocationLookUp` map feels not scalable to me if we want to make it a useful generic implementation. For example, customers might want to have 1 HDFS cluster per [AWS local zone](https://aws.amazon.com/about-aws/global-infrastructure/localzones/) to read and write data due to latency requirement, and this setup easily generate hundreds to thousands of table properties. I don't know if there is any bad impact for such use case, since table properties is not designed to hold a lot of data.

mickjermsurawong-stripe commented:
hi @jackye1995 thanks for the feedback here! This is aligned with the discussion we had during the sync for solution that folks can easily customizable their locations. I've re-purposed this PR to address this. If this is in, we can easily implement our data localization here.

jackye1995 commented:
Thank you! I created an issue for this, can you link it to this PR? #1538

mickjermsurawong-stripe commented:
cc @stripe/iceberg-reviewers, I iterated on this based on feedback to have pluggable impl for location provider instead

rdblue commented:
Thanks for working on this, @mickjermsurawong-stripe!

I like dynamically loading a location provider, and I like that the changes required are pretty small. That looks good to me.

What I'd like to change is how this is tested. There are two main issues with it. First, the tests are in an unrelated suite for Spark that is mainly testing that Spark works with a variety of different schemas. It doesn't seem appropriate to put the tests for injecting a location provider in a different module and into this suite. Second, the `TestLocationProvider` returns real paths and the tests work by writing data files and validating that files are underneath those paths. That does test that it works, but I think this makes validation much harder. This approach requires a lot of test code and runs a lot of unrelated cases through to validate the location provider.

I would prefer a more targeted test in a new suited in core, `TestLocationProvider`. That would have a case that injects a `TestLocationProvider` that keeps track of calls in static variables, so you'd be able to inject the provider, get it from `table.operations().locationProvider()`, call some methods, and validate that `TestLocationProvider` received the calls.

We also need negative tests for when the provider can't be found. I'd like to have a good error message like "Cannot find location provider class: %s".

rdblue commented:
@jackye1995, I think this dynamic loading approach might be a good idea for injecting S3FileIO as well. What do you think?

jackye1995 commented:
@rdblue yes, that is exactly how I implement, just waiting for your branch to cross check all details about S3FileIO before opening a PR

mickjermsurawong-stripe commented:
@rdblue, ah thank you for the feedback on test! I guess the complicated test here is the artifacts of me trying to test the logic in my first data localization feature. 

I created a separate test as suggested; I made use of the existing `TableTestBase` and construct tests based on `table.locationProvider()` instead of its underlying operations (though that also directly delegates to the ops). I also addressed feedback on error message.

rdblue commented on core/src/test/java/org/apache/iceberg/TestLocationProvider.java:
```diff
@@ -0,0 +1,164 @@
+
+  // publicly visible for testing to be dynamically loaded
+  public static class InvalidDynamicallyLoadedLocationProvider implements LocationProvider {
+    // No public constructor
```
Should we support a no-arg constructor as well as the one that passes the table location and properties?

rdblue commented on core/src/test/java/org/apache/iceberg/TestLocationProvider.java:
```diff
@@ -0,0 +1,164 @@
+  }
+
+  @Test
+  public void testInvalidDynamicallyLoadedLocationProvider() {
```
What happens if the class doesn't implement `LocationProvider`?

Could you add a test case for that?

aokolnychyi commented on core/src/main/java/org/apache/iceberg/TableProperties.java:
```diff
@@ -87,6 +87,8 @@ private TableProperties() {
   public static final String OBJECT_STORE_ENABLED = "write.object-storage.enabled";
   public static final boolean OBJECT_STORE_ENABLED_DEFAULT = false;
 
+  public static final String LOCATION_PROVIDER_IMPL = "write.location-provider.impl";
```
Do we want to add this to the docs?

rdblue replied:
Yes

rdblue commented on core/src/main/java/org/apache/iceberg/LocationProviders.java:
```diff
@@ -45,6 +50,53 @@ public static LocationProvider locationsFor(String location, Map<String, String>
+        DynConstructors.builder(LocationProvider.class)
+            .impl(impl).build()
+    );
+    Optional<DynConstructors.Ctor<LocationProvider>> twoArgConstructor = Optional.empty();
```
When the arguments passed to one constructor are a prefix of the args passed to the other, there is no need to use a different builder or call because the arguments actually passed are limited by the number of parameters defined by the constructor. That makes it possible to pass both arguments to either one, and any extras are ignored.

That helps simplify cases like this because you don't need the extra optional logic:

```java
  ctor = DynConstructors.builder(LocationProvider.class)
      .impl(impl, String.class, Map.class)
      .impl(impl) // fall back to no-arg constructor
      .build()
  return ctor.newInstance(location, properties);
```

mickjermsurawong-stripe replied:
thanks for the explanation here

rdblue commented on site/docs/javadoc/master/constant-values.html:
```diff
@@ -535,13 +535,20 @@ <h2 title="org.apache">org.apache.*</h2>
 <td class="colLast"><code>"write.metadata.path"</code></td>
 </tr>
-<tr class="altColor">
+<tr class="rowColor">
```
Can you remove the html changes please?

mickjermsurawong-stripe replied:
ahh. ack!

rdblue commented:
Thanks for the update, @mickjermsurawong-stripe! I think the implementation can be quite a bit simpler and more like the original, even with support for the no-arg constructor. Sorry I didn't mention that earlier, before you went ahead and implemented it this way.

mickjermsurawong-stripe commented on core/src/main/java/org/apache/iceberg/LocationProviders.java:
```diff
@@ -36,7 +37,29 @@ private LocationProviders() {
+                "has a public no-arg constructor or a two-arg constructor " +
+                "taking in the string base table location and its property string map.",
+            impl, LocationProvider.class));
+      }
```
If provided class has the specified constructor signature, but not implementing `LocationProvider`, it can still be loaded into the dynConstructor. The following block differentiates this missing interface more explicitly. 
Might be a bit more code churn, but it does give more signal to users.

mickjermsurawong-stripe commented:
thanks @rdblue for the review so far! I believe this should be ready for another review.

rdblue commented on core/src/main/java/org/apache/iceberg/LocationProviders.java:
```diff
@@ -36,7 +37,29 @@ private LocationProviders() {
+      } catch (ClassCastException e) {
+        throw new IllegalArgumentException(
+            String.format("Provided implementation for dynamic instantiation should implement %s, " +
+                "but found dynamic constructor %s.", LocationProvider.class, ctor));
```
Dynamic constructor?

rdblue replied:
Here as well you should probably throw the exception with the original as a cause so the context isn't lost.

rdblue commented on core/src/main/java/org/apache/iceberg/LocationProviders.java:
```diff
@@ -36,7 +37,29 @@ private LocationProviders() {
+                "Make sure the implementation is in classpath, and that it either " +
+                "has a public no-arg constructor or a two-arg constructor " +
+                "taking in the string base table location and its property string map.",
+            impl, LocationProvider.class));
```
Since you're catching the exception, it probably makes more sense to use `buildChecked` so the exception class isn't generic. That exception also has good information about why all of the implementations failed, so I would recommend adding that exception as a cause to this one that is thrown. I like throwing this one for context, though!

mickjermsurawong-stripe commented:
Propagating the causes now should give stacktrace for these two cases:

```
 java.lang.IllegalArgumentException: Unable to find a constructor for implementation org.apache.iceberg.TestLocationProvider$InvalidArgTypesDynamicallyLoadedLocationProvider of interface org.apache.iceberg.io.LocationProvider. Make sure the implementation is in classpath, and that it either has a public no-arg constructor or a two-arg constructor taking in the string base table location and its property string map.
        at org.apache.iceberg.LocationProviders.locationsFor(LocationProviders.java:48)
        at org.apache.iceberg.TestTables$TestTableOperations.locationProvider(TestTables.java:212)
        at org.apache.iceberg.BaseTable.locationProvider(BaseTable.java:199)
        at org.apache.iceberg.TestLocationProvider.testInvalidArgTypesDynamicallyLoadedLocationProvider(TestLocationProvider.java:208)

        Caused by:
        java.lang.NoSuchMethodException: Cannot find constructor for interface org.apache.iceberg.io.LocationProvider
                Missing org.apache.iceberg.TestLocationProvider$InvalidArgTypesDynamicallyLoadedLocationProvider() [java.lang.NoSuchMethodException: org.apache.iceberg.TestLocationProvider$InvalidArgTypesDynamicallyLoadedLocationProvider.<init>()]
                Missing org.apache.iceberg.TestLocationProvider$InvalidArgTypesDynamicallyLoadedLocationProvider(java.lang.String,java.util.Map) [java.lang.NoSuchMethodException: org.apache.iceberg.TestLocationProvider$InvalidArgTypesDynamicallyLoadedLocationProvider.<init>(java.lang.String, java.util.Map)]
```

and

```
java.lang.IllegalArgumentException: Provided implementation for dynamic instantiation should implement interface org.apache.iceberg.io.LocationProvider.
        at org.apache.iceberg.LocationProviders.locationsFor(LocationProviders.java:59)
        at org.apache.iceberg.TestTables$TestTableOperations.locationProvider(TestTables.java:212)
        at org.apache.iceberg.BaseTable.locationProvider(BaseTable.java:199)
        at org.apache.iceberg.TestLocationProvider.testInvalidNoInterfaceDynamicallyLoadedLocationProvider(TestLocationProvider.java:190)

        Caused by:
        java.lang.ClassCastException: org.apache.iceberg.TestLocationProvider$InvalidNoInterfaceDynamicallyLoadedLocationProvider cannot be cast to org.apache.iceberg.io.LocationProvider
            at org.apache.iceberg.LocationProviders.locationsFor(LocationProviders.java:56)
            ... 3 more
```

rdblue commented:
Thanks, @mickjermsurawong-stripe! I merged this. Nice work, and thank you for your patience and quick turn-around with the review comments!

mickjermsurawong-stripe commented:
Thank _you_ Ryan! This will help with our internal feature :)
